spec/watirspec/selector_builder/cell_spec.rb

```diff
@@ -0,0 +1,57 @@
+require 'watirspec_helper'
+
+describe Watir::Locators::Cell::SelectorBuilder do
+  let(:attributes) { Watir::HTMLElement.attribute_list }
+  let(:selector_builder) { described_class.new(attributes) }
+
+  describe '#build' do
+    after(:each) do |example|
+      next if example.metadata[:skip_after]
+
+      @query_scope ||= browser.element(id: 'gregory').locate
+
+      built = selector_builder.build(@selector)
+      expect(built).to eq [@wd_locator, (@remaining || {})]
+
+      next unless @data_locator || @tag_name
+
+      expect { @located = @query_scope.wd.first(@wd_locator) }.not_to raise_exception
+
+      if @data_locator
+        expect(@located.attribute('data-locator')).to eq(@data_locator)
+      else
+        expect {
+          expect(@located.tag_name).to eq @tag_name
+        }.not_to raise_exception
+      end
+    end
+
+    it 'without any arguments' do
+      browser.goto(WatirSpec.url_for('tables.html'))
+      @selector = {}
+      @wd_locator = {xpath: "./*[local-name()='th' or local-name()='td']"}
+      @data_locator = 'first cell'
+    end
+
+    context 'with multiple locators' do
+      before(:each) do
+        browser.goto(WatirSpec.url_for('tables.html'))
+      end
+
+      it 'attribute and text' do
+        @selector = {headers: /before_tax/, text: '5 934'}
+        @wd_locator = {xpath: "./*[local-name()='th' or local-name()='td']" \
+"[contains(@headers, 'before_tax')][normalize-space()='5 934']"}
+        @data_locator = 'before tax'
+      end
+    end
+
+    it 'delegates adjacent to Element SelectorBuilder' do
```

Comment:
Isn't this saying:

~~~~~~~~
browser.table.tr.cell(adjacent: :ancestor, index: 2)
#=> Watir::Table
~~~~~~~~

Given we are call `#cell`, wouldn't one expect to find a `th` or `td`? "cell" is supposed to be the immediate child of a given row, so does that make it weird to be going even further up? I haven't used :adjacent directly, so maybe I'm thinking of this wrong?

Reply:
No, you're right. I actually spent several hours working with adjacent specs and implementation ideas, until I had the "I'm an idiot moment" and realized what I was doing didn't make sense. All this spec is doing is code coverage to demonstrate that return super line is hit in the cell XPath class for adjacent tags.

Reply:
I'd be tempted to just throw an exception if someone specified :adjacent. I think it'd allow you to simplify such that you don't have to manually decomponse and recompose the xpath. This re-building of the xpath was quite confusing to me. I would have thought that you could just re-define certain methods and #build would just work.

Reply:
Well, I guess I'm reluctant to throw an exception when it is handled easily enough....
Something like?

```ruby
table.row(id: 'outer_last').child(index: -1)
```
`#row` matches any `tr` under `tbody`, `thead` or `tbody`, though now that I think about it, I'm not sure why we are jumping through those hoops, because `tr` by itself should do that just fine... I think we can investigate more optimization there...

Actually, maybe I should remove all index matching with adjacent locators in this PR, and then add back in a global solution in a follow-up...

Reply:
Really these special SelectorBuilder classes are only used internally by Table & Row subclasses to support things like `#hashes` and `#headers`. Also, we leverage them to do `attribute_value(:rowIndex)` because the order that is in the DOM is not the order of what is displayed on the page.

Reply:
`#row` and `#tr` are not the same:
* `#row` matches only `tr` elements in the current table (ignores nested tables)
` #tr` matches all descendant `tr`, including those in nested tables

Reply:
I'm not sure I understand the intent of the `table.row(id: 'outer_last').child(index: -1)` example. That wouldn't end up calling the cell selector builder?

Reply:
No, it will make an adjacent call on the Row builder.
and `child` will work with both `th` and `td` elements.

This isn't a great example, but I think it should be a valid one, and it isn't that difficult to support with this code.